File: fakenewsdetaction-main/serve-ui.py
#!/usr/bin/env python3
"""
Simple HTTP Server for Fake News Detection UI
Serves the frontend files on http://localhost:3000
"""

import http.server
import socketserver
import os
import sys
from pathlib import Path

# Configuration
PORT = 3000
DIRECTORY = "ui"

class CustomHTTPRequestHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=DIRECTORY, **kwargs)
    
    def end_headers(self):
        # Add CORS headers
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        super().end_headers()
    
    def log_message(self, format, *args):
        # Custom logging format
        print(f"[Frontend] {self.address_string()} - {format % args}")

def main():
    # Check if ui directory exists
    ui_path = Path(DIRECTORY)
    if not ui_path.exists():
        print(f"❌ Error: '{DIRECTORY}' directory not found!")
        print("Make sure you're running this from the project root directory.")
        sys.exit(1)
    
    # Check if index.html exists
    index_path = ui_path / "index.html"
    if not index_path.exists():
        print(f"❌ Error: 'index.html' not found in '{DIRECTORY}' directory!")
        sys.exit(1)
    
    print("🌐 Starting Frontend Server...")
    print(f"📁 Serving files from: {ui_path.absolute()}")
    print(f"🔗 Frontend URL: http://localhost:{PORT}")
    print(f"🔗 API Backend: http://localhost:8080")
    print("🛑 Press Ctrl+C to stop the server")
    print()
    
    try:
        with socketserver.TCPServer(("", PORT), CustomHTTPRequestHandler) as httpd:
            print(f"✅ Frontend server started on http://localhost:{PORT}")
            print("📝 Open http://localhost:3000 in your browser")
            httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n🛑 Frontend server stopped")
    except OSError as e:
        if e.errno == 10048:  # Port already in use on Windows
            print(f"❌ Error: Port {PORT} is already in use!")
            print("Try stopping other servers or use a different port.")
        else:
            print(f"❌ Error starting server: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: fakenewsdetaction-main/serve-ui.js
#!/usr/bin/env node

/**
 * Simple Node.js HTTP Server for Fake News Detection UI
 * Alternative to Python server - serves frontend on http://localhost:3000
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const url = require('url');

// Configuration
const PORT = process.env.PORT || 3000;
const UI_DIR = 'ui';

// MIME types for different file extensions
const MIME_TYPES = {
    '.html': 'text/html',
    '.css': 'text/css',
    '.js': 'application/javascript',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.ico': 'image/x-icon'
};

function getMimeType(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    return MIME_TYPES[ext] || 'text/plain';
}

function serveFile(res, filePath) {
    fs.readFile(filePath, (err, data) => {
        if (err) {
            res.writeHead(404, { 'Content-Type': 'text/plain' });
            res.end('404 - File Not Found');
            return;
        }

        const mimeType = getMimeType(filePath);
        res.writeHead(200, {
            'Content-Type': mimeType,
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type'
        });
        res.end(data);
    });
}

function handleRequest(req, res) {
    const parsedUrl = url.parse(req.url, true);
    let pathname = parsedUrl.pathname;

    // Default to index.html for root path
    if (pathname === '/') {
        pathname = '/index.html';
    }

    // Remove leading slash and resolve file path
    const filePath = path.join(UI_DIR, pathname.substring(1));
    
    // Security check - prevent directory traversal
    const resolvedPath = path.resolve(filePath);
    const resolvedUIDir = path.resolve(UI_DIR);
    
    if (!resolvedPath.startsWith(resolvedUIDir)) {
        res.writeHead(403, { 'Content-Type': 'text/plain' });
        res.end('403 - Forbidden');
        return;
    }

    // Log the request
    const timestamp = new Date().toLocaleTimeString();
    console.log(`[${timestamp}] ${req.method} ${req.url} -> ${filePath}`);

    // Check if file exists
    fs.stat(resolvedPath, (err, stats) => {
        if (err) {
            res.writeHead(404, { 'Content-Type': 'text/plain' });
            res.end('404 - File Not Found');
            return;
        }

        if (stats.isFile()) {
            serveFile(res, resolvedPath);
        } else {
            res.writeHead(404, { 'Content-Type': 'text/plain' });
            res.end('404 - Not a file');
        }
    });
}

function main() {
    // Check if UI directory exists
    if (!fs.existsSync(UI_DIR)) {
        console.error(`❌ Error: '${UI_DIR}' directory not found!`);
        console.error('Make sure you\'re running this from the project root directory.');
        process.exit(1);
    }

    // Check if index.html exists
    const indexPath = path.join(UI_DIR, 'index.html');
    if (!fs.existsSync(indexPath)) {
        console.error(`❌ Error: 'index.html' not found in '${UI_DIR}' directory!`);
        process.exit(1);
    }

    console.log('🌐 Starting Frontend Server...');
    console.log(`📁 Serving files from: ${path.resolve(UI_DIR)}`);
    console.log(`🔗 Frontend URL: http://localhost:${PORT}`);
    console.log(`🔗 API Backend: http://localhost:8080`);
    console.log('🛑 Press Ctrl+C to stop the server');
    console.log('');

    // Create HTTP server
    const server = http.createServer(handleRequest);

    // Handle server errors
    server.on('error', (err) => {
        if (err.code === 'EADDRINUSE') {
            console.error(`❌ Error: Port ${PORT} is already in use!`);
            console.error('Try stopping other servers or use a different port.');
        } else {
            console.error(`❌ Error starting server: ${err.message}`);
        }
        process.exit(1);
    });

    // Start the server
    server.listen(PORT, () => {
        console.log(`✅ Frontend server started on http://localhost:${PORT}`);
        console.log(`📝 Open http://localhost:${PORT} in your browser`);
    });

    // Handle graceful shutdown
    process.on('SIGINT', () => {
        console.log('\n🛑 Frontend server stopped');
        server.close(() => {
            process.exit(0);
        });
    });

    process.on('SIGTERM', () => {
        console.log('\n🛑 Frontend server stopped');
        server.close(() => {
            process.exit(0);
        });
    });
}

if (require.main === module) {
    main();
}